### PR TITLE
Patch 5154 Undertaker Test

### PR DESCRIPTION
**Pull request by ThePhisch**
Extension of the Undertaker test suite and changes to the underlying rucio codebase in order to ensure that the tests pass. Includes the following testcases:

`TestUndertaker::test_undertaker` in which files belonging to expired datasets are tombstoned and the datasets deleted. The image only shows one dataset with three files, the actual test has many more datasets and larger amounts of files attached to each. The green diamond denotes a rule protecting the dataset.
test_undertaker

`test_file_protected_by_rule` in which a file (F1) belonging to an expired dataset is protected by an external second rule (R2). The protected file is detached from the expired dataset (dotted line). The unprotected file (F2) is also detached and its replicas are tombstoned. The F1 replicas protected by R2 remain untouched, whereas the F1 replicas indirectly protected by R1 via the dataset are tombstoned.
test_file_protected_by_rule

`test_file_protected_by_dataset`: similarly, F1 belongs to a nonexpired dataset D2 as well as the expired dataset D1.
test_file_protected_by_dataset

`test_file_dids_in_dataset` covers the situation in which an expired file DID (F3) is attached to a nonexpired dataset. The expected outcome is that F3 is detached and its replicas are tombstoned.
test_file_dids_in_dataset

`test_file_did_deletion` checks whether a file F that is _not_ attached to any dataset has its replicas tombstoned. The rule R is optional in the actual test (the test runs once with and once without the rule).
test_file_did_deletion

In summary, this PR ensures that the undertaker fulfills its mission of handling expired DIDs properly. In general, an expired DID has all its replicas (be they directly associated because the expired DID is a file or indirectly because the expired DID is a collection) receive the epoch-tombstone (a special datetime marker set in 1970). In addition, any attached DIDs are detached. The behaviour becomes more complicated when additional rules and attachments (to a parent) are at play. Additional rules and parents protect any non-expired DIDs which would otherwise be affected by a parent being expired.

The case for containers is currently being argued in #6055

### CODE REVIEW

File: lib/rucio/core/did.py [hunk 1]
```diff
@@ -1650,6 +1666,20 @@ def _delete_dids_wo_temp_tables(
...
+        query = update(
+            models.RSEFileAssociation
+        ).filter(
+            models.RSEFileAssociation.scope == scope,
+            models.RSEFileAssociation.name == name
+        ).values(
+            tombstone=OBSOLETE
+        )
+        session.execute(query)
+        pass
```
Comment (mlassnig) [inline lib/rucio/core/did.py]: unnecessary pass?

File: lib/rucio/core/did.py [hunk 2]
```diff
@@ -1290,14 +1290,28 @@ def _delete_dids(
...
+                query = update(
+                    models.RSEFileAssociation
+                ).filter(
+                    models.RSEFileAssociation.scope == didscope,
+                    models.RSEFileAssociation.name == didname
+                ).values(
+                    tombstone=OBSOLETE
+                )
+                session.execute(query)
+                continue
```
Comment (mlassnig) [inline lib/rucio/core/did.py:1315]: unnecessary continue? or is it just to skip the logger?

File: lib/rucio/tests/test_undertaker.py [hunk 3]
```diff
@@ -15,22 +15,33 @@
 
 from datetime import datetime, timedelta
 from logging import getLogger
+from collections import namedtuple
```
Comment (mlassnig) [inline lib/rucio/tests/test_undertaker.py]: sort alphabetically

File: lib/rucio/tests/test_undertaker.py [hunk 4]
```diff
@@ -15,22 +15,33 @@
...
+from rucio.db.sqla.constants import OBSOLETE
 from rucio.tests.common import rse_name_generator, did_name_generator
+from rucio.common.exception import DataIdentifierNotFound
+
+from typing import TYPE_CHECKING
+if TYPE_CHECKING:
+    from rucio.tests.temp_factories import TemporaryRSEFactory, TemporaryDidFactory
 
 LOG = getLogger(__name__)
+Rse = namedtuple('Rse', ['name', 'id'])
```
Comment (mlassnig) [inline lib/rucio/tests/test_undertaker.py]: `Rse`  capitalisation looks... very wrong

File: lib/rucio/core/did.py (hunk 1 above)
Comment (ThePhisch) [reply lib/rucio/core/did.py]: yup, removed it

File: lib/rucio/core/did.py (hunk 2 above)
Comment (ThePhisch) [reply lib/rucio/core/did.py:1315]: necessary continue: this `if` block is reached when no rule could be queried for the given scopename. Some calls in the following code (i.e. `rucio.core.rule.delete_rule(rule_id=rule_id, ...`) will lead to errors when run with `rule_id = None`.

The point you made with the logger is a good one, I added logger calls.

File: lib/rucio/tests/test_undertaker.py (hunk 4 above)
Comment (ThePhisch) [reply lib/rucio/tests/test_undertaker.py]: true. Changed it to `RSE_namedtuple`

File: lib/rucio/core/rule.py [hunk 5]
```diff
@@ -889,6 +889,32 @@ def delete_rule(rule_id, purge_replicas=None, soft=False, delete_parent=False, n
         transfers_to_delete = []  # [{'scope': , 'name':, 'rse_id':}]
         account_counter_decreases = {}  # {'rse_id': [file_size, file_size, file_size]}
 
+        if locks.first() is None:
```
Comment (bari12) [inline lib/rucio/core/rule.py:892]: Why is this entire part needed?

File: lib/rucio/core/did.py [hunk 6]
```diff
@@ -1290,14 +1290,29 @@ def _delete_dids(
             models.ReplicationRule.rse_expression,
             models.ReplicationRule.locks_ok_cnt,
             models.ReplicationRule.locks_replicating_cnt,
-            models.ReplicationRule.locks_stuck_cnt
-        ).join_from(
+            models.ReplicationRule.locks_stuck_cnt,
+            temp_table.scope,
+            temp_table.name
+        ).outerjoin_from(
```
Comment (bari12) [inline lib/rucio/core/did.py:1296]: Why is this an outer join now?

**Review by bari12 — COMMENTED** [review]
I think the tests are good, but I don't fully follow some of the did/rule changes. @cserf please have a look on this as well.
